gunrock/app/bc/bc_enactor.cuh

```diff
@@ -14,2346 +14,645 @@
-    typename SizeT,
-    typename Value>
-__global__ void Expand_Incoming_Backward_Kernel (
-    const SizeT           num_elements,
-    const VertexId* const d_keys_in,
-    const Value*    const d_values_in,
-          SizeT           output_offset,
-          VertexId*       d_keys_out,
-          Value*          d_deltas,
-          Value*          d_bc_values,
-          VertexId*       d_original_vertex)
-{
-    SizeT x = (SizeT) blockIdx.x * blockDim.x + threadIdx.x;
-    const SizeT STRIDE = (SizeT) gridDim.x * blockDim.x;
-    while (x < num_elements)
-    {
-        VertexId key = d_keys_in[x];
-        d_deltas   [key] = d_values_in[x << 1];
-        d_bc_values[key] = d_values_in[(x << 1) + 1];
-        //d_keys_out[x + output_offset] = key;
-        //printf("Expand_Incoming : [%2d, %2d] -> %.4f, %.4f\n",
-        //    x, d_original_vertex[key], d_values_in[x << 1], d_values_in[(x<<1)+1]);
-        x += STRIDE;
-    }
-}
+    BCForwardIterationLoop() : BaseIterationLoop() {}
 
-template <
-    typename VertexId,
-    typename SizeT,
-    typename Value>
-__global__ void Assign_Middle(
-    int        num_gpus,
-    SizeT      num_elements,
-    VertexId  *d_keys_in,
-    VertexId  *d_labels_in,
-    Value     *d_sigmas_in,
-    //VertexId **d_keys_outs,
-    VertexId  *d_labels_out,
-    Value     *d_sigmas_out)
-{
-    SizeT x = (SizeT)blockIdx.x * blockDim.x + threadIdx.x;
-    const SizeT STRIDE = (SizeT) gridDim.x * blockDim.x;
-    /*__shared__ VertexId* s_keys_outs[8];
-    __shared__ VertexId* s_labels_outs[8];
-    __shared__ Value*    s_sigmas_outs[8];
-    if (threadIdx.x < num_gpus)
-    {
-        s_keys_outs  [threadIdx.x] = d_keys_outs  [threadIdx.x];
-        s_labels_outs[threadIdx.x] = d_labels_outs[threadIdx.x];
-        s_sigmas_outs[threadIdx.x] = d_sigmas_outs[threadIdx.x];
-    }
-    __syncthreads();*/
+    /**
+     * @brief Core computation of bc, one iteration
+     * @param[in] peer_ Which GPU peers to work on, 0 means local
+     * \return cudaError_t error message(s), if any
+     */
+    cudaError_t Core(int peer_ = 0)
+    {
+        // Data alias the enactor works on
+        auto    &data_slice         =   this -> enactor ->
+            problem -> data_slices[this -> gpu_num][0];
+        auto    &enactor_slice      =   this -> enactor ->
+            enactor_slices[this -> gpu_num * this -> enactor -> num_gpus + peer_];
+        auto    &enactor_stats      =   enactor_slice.enactor_stats;
+        auto    &graph              =   data_slice.sub_graph[0];
+        auto    &frontier           =   enactor_slice.frontier;
+        auto    &oprtr_parameters   =   enactor_slice.oprtr_parameters;
+        auto    &retval             =   enactor_stats.retval;
+
+        // BC specific data alias here, e.g.:
+        auto    &sigmas             =   data_slice.sigmas;
+        auto    &labels             =   data_slice.labels;
+
+        // ----------------------------
+        // Forward advance
```

Comment:
I think it is worth explaining what happens in this advance and filter for forward iteration and backward iteration in the comments. Essentially the first pass is a BFS and second is calculating the bc values.

Reply:
added a note